### ISSUE TITLE
Correctness issue in boolean expression simplification

### ISSUE THREAD

**Issue by haozhun** [prestodb/presto#4546]
This query should produce one row. It produces two rows with #4431. #4431 was backed out in #4545.
```
  SELECT
    *
  FROM (
    values
      ('1234', 'a', '2016-01-01'),
      ('1235', 'b', '2016-01-01'),
  ) z(app, s, ds)
  WHERE
    ds > '2015-01-01' AND
    (app like '%234' OR app like '%235' OR
    app like '234%' OR app like '235%') AND
    (app like '%235' OR app like '235%')
```

**Comment by haozhun** [comment]
@sopel39 Please take a look.

**Comment by sopel39** [comment]
Indeed there was a problem when one predicate was subset of another, e.g:
`(A AND B) OR (A AND B AND C)`
or
`(A OR B) AND (A OR B OR C)`

In such case the rewrite should proceed:
`(A AND B) OR (A AND B AND C) => (A AND B) AND ((TRUE) OR (C)) => (A AND B)`
and
`(A OR B) AND (A OR B OR C) => (A OR B) OR ((FALSE) AND (C)) => (A OR B)`
This was fixed in #4551.

Could you give us more information how did you found this problem? Was it using `presto-verifier` or some production (nope not) environment?

**Comment by electrum** [comment]
We found this in the verifier.
